internal/orchestrator/reorg_handler.go

```diff
@@ -0,0 +1,187 @@
+				continue
+			}
+			metrics.ReorgCounter.Inc()
+			forkPoint, err := rh.findForkPoint(blockHeaders[reorgEndIndex:])
+			if err != nil {
+				log.Error().Err(err).Msg("Error while finding fork point")
+				continue
+			}
+			err = rh.handleReorg(forkPoint, lookbackFrom)
+			if err != nil {
+				log.Error().Err(err).Msg("Error while handling reorg")
+				continue
+			}
+			rh.lastCheckedBlock = lookbackFrom
+			rh.storage.OrchestratorStorage.SetLastReorgCheckedBlockNumber(rh.rpc.ChainID, lookbackFrom)
+			metrics.ReorgHandlerLastCheckedBlock.Set(float64(lookbackFrom.Int64()))
+		}
+	}()
+
+	// Keep the program running (otherwise it will exit)
+	select {}
+}
+
+func findReorgEndIndex(reversedBlockHeaders []common.BlockHeader) (index int) {
+	for i := 0; i < len(reversedBlockHeaders)-1; i++ {
+		currentBlock := reversedBlockHeaders[i]
+		previousBlock := reversedBlockHeaders[i+1]
+
+		if currentBlock.ParentHash != previousBlock.Hash {
+			log.Info().
+				Str("currentBlockNumber", currentBlock.Number.String()).
+				Str("currentBlockHash", currentBlock.Hash).
+				Str("currentBlockParentHash", currentBlock.ParentHash).
+				Str("previousBlockNumber", previousBlock.Number.String()).
+				Str("previousBlockHash", previousBlock.Hash).
+				Msg("Reorg detected: parent hash mismatch")
+			return i
+		}
+	}
+	return -1
+}
+
+func (rh *ReorgHandler) findForkPoint(reversedBlockHeaders []common.BlockHeader) (forkPoint *big.Int, err error) {
+	blockNumbers := make([]*big.Int, 0, len(reversedBlockHeaders))
+	for _, header := range reversedBlockHeaders {
+		blockNumbers = append(blockNumbers, header.Number)
+	}
+	blockResults := rh.rpc.GetBlocks(blockNumbers)
+	fetchedBlocksByNumber := make(map[string]common.Block)
+	for _, blockResult := range blockResults {
+		if blockResult.Error != nil {
+			return nil, fmt.Errorf("error fetching block %s: %w", blockResult.BlockNumber.String(), blockResult.Error)
+		}
+		fetchedBlocksByNumber[blockResult.BlockNumber.String()] = blockResult.Data
+	}
+
+	for i := 0; i < len(reversedBlockHeaders)-1; i++ {
+		blockHeader := reversedBlockHeaders[i]
+		block, ok := fetchedBlocksByNumber[blockHeader.Number.String()]
+		if !ok {
+			return nil, fmt.Errorf("block not found: %s", blockHeader.Number.String())
+		}
+		if block.Hash == blockHeader.Hash {
+			previousBlock := reversedBlockHeaders[i+1]
+			return previousBlock.Number, nil
+		}
+	}
+	lookbackFrom := reversedBlockHeaders[len(reversedBlockHeaders)-1].Number
+	nextHeadersBatch, err := rh.storage.MainStorage.LookbackBlockHeaders(rh.rpc.ChainID, rh.blocksPerScan, lookbackFrom)
+	if err != nil {
+		return nil, fmt.Errorf("error getting next headers batch: %w", err)
+	}
+	return rh.findForkPoint(nextHeadersBatch)
+}
+
+func (rh *ReorgHandler) handleReorg(reorgStart *big.Int, reorgEnd *big.Int) error {
+	blockRange := make([]*big.Int, 0)
+	for i := new(big.Int).Set(reorgStart); i.Cmp(reorgEnd) <= 0; i.Add(i, big.NewInt(1)) {
```

Comment:
why big int inside the for definition and big int casting again inside the loop (L163)?

Reply:
i will change as the loop progresses, we want to create a new big int